### PR TITLE
Discard minimum gp tx and optimizate mempool

### PR DESCRIPTION
**Pull request by BananaLF**
update deley logic
pending tx can not delete
delete tx in update
solve a bug for can not allow to checktx uesd concurrency when mempool is full. use FlagMempoolEnableDeleteMinGPTx to control thie feature finsh pengdingPool may be block main goroutine and occupy too much system resources at the extreme condition discard minimum gp tx and optimizate mempool

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: dev/start.sh [hunk 1]
```diff
@@ -22,13 +22,14 @@ killbyname() {
...
     LOG_LEVEL=main:info,iavl:info,*:error,state:info,provider:info
-
+#--mempool.enable_delete_min_gp_tx false \
 #    exchaind start --pruning=nothing --rpc.unsafe \
     exchaind start --rpc.unsafe \
       --local-rpc-port 26657 \
       --log_level $LOG_LEVEL \
       --log_file json \
       --dynamic-gp-mode=2 \
+      --mempool.enable_delete_min_gp_tx=true \
```
Comment (xzav3r) [inline dev/start.sh]: it's better to set false for `mempool.enable_delete_min_gp_tx` in default

File: libs/tendermint/config/dynamic_config_okchain.go [hunk 2]
```diff
@@ -116,3 +118,11 @@ func (d MockDynamicConfig) GetDeliverTxsExecuteMode() int {
 func (d MockDynamicConfig) GetEnableHasBlockPartMsg() bool {
 	return false
 }
+
+func (d MockDynamicConfig) GetEnableDeleteMinGPTx() bool {
+	return d.enableDeleteMinGPTx
+}
+
+func (d *MockDynamicConfig) SetEnableDeleteMinGPTx(enable bool) {
+	d.enableDeleteMinGPTx = true
```
Comment (xzav3r) [inline libs/tendermint/config/dynamic_config_okchain.go]: d.enableDeleteMinGPTx = enable

File: libs/tendermint/mempool/clist_mempool.go [hunk 3]
```diff
@@ -482,7 +497,19 @@ func (mem *CListMempool) addPendingTx(memTx *mempoolTx) error {
 	if txNonce == expectedNonce {
 		err := mem.addTx(memTx)
 		if err == nil {
-			go mem.consumePendingTx(memTx.from, txNonce+1)
+			if len(mem.consumePendingTxQueue) >= mem.consumePendingTxQueueLimit {
```
Comment (xzav3r) [inline libs/tendermint/mempool/clist_mempool.go]: use `select` instead of `if`

File: libs/tendermint/mempool/clist_mempool.go [hunk 4]
```diff
@@ -964,8 +1005,15 @@ func (mem *CListMempool) Update(
 	// Update metrics
 	mem.metrics.Size.Set(float64(mem.Size()))
 	if mem.pendingPool != nil {
-		mem.pendingPoolNotify <- addressNonce
-		mem.metrics.PendingPoolSize.Set(float64(mem.pendingPool.Size()))
+		if len(mem.pendingPoolNotify) >= 1 {
```
Comment (xzav3r) [inline libs/tendermint/mempool/clist_mempool.go]: use `select` instead of `if `

**Review by xzav3r — DISMISSED** [review]

File: libs/tendermint/cmd/tendermint/commands/run_node.go [hunk 5]
```diff
@@ -116,6 +116,11 @@ func AddNodeFlags(cmd *cobra.Command) {
 		config.Mempool.MaxTxNumPerBlock,
 		"Maximum number of transactions in a block",
 	)
+	cmd.Flags().Bool(
+		"mempool.enable_delete_min_gp_tx",
+		config.Mempool.EnableDeleteMinGPTx,
+		"Enable delete the minimum gas price tx from mempool",
```
Comment (ilovers) [inline libs/tendermint/cmd/tendermint/commands/run_node.go]:
```suggestion
		"Enable delete the minimum gas price tx from mempool when mempool is full",
```

**Review by ilovers — DISMISSED** [review]

**Review by ilovers — DISMISSED** [review]

**Review by xzav3r — DISMISSED** [review]

**Review by ilovers — APPROVED** [review]

**Review by xzav3r — APPROVED** [review]

**Pull request merged**